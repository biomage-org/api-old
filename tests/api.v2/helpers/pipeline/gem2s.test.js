// @ts-nocheck
const io = require('socket.io-client');

const { createGem2sPipeline, handleGem2sResponse } = require('../../../../src/api.v2/helpers/pipeline/gem2s');

const Experiment = require('../../../../src/api.v2/model/Experiment');
const Sample = require('../../../../src/api.v2/model/Sample');
const ExperimentExecution = require('../../../../src/api.v2/model/ExperimentExecution');

const pipelineConstruct = require('../../../../src/api.v2/helpers/pipeline/pipelineConstruct');
const getPipelineStatus = require('../../../../src/api.v2/helpers/pipeline/getPipelineStatus');
const HookRunner = require('../../../../src/api.v2/helpers/pipeline/hooks/HookRunner');

const validateRequest = require('../../../../src/utils/schema-validator');

const constants = require('../../../../src/api.v2/constants');

jest.mock('socket.io-client');

jest.mock('../../../../src/api.v2/model/Experiment');
jest.mock('../../../../src/api.v2/model/Sample');
jest.mock('../../../../src/api.v2/model/ExperimentExecution');

jest.mock('../../../../src/api.v2/helpers/pipeline/pipelineConstruct');
jest.mock('../../../../src/api.v2/helpers/pipeline/getPipelineStatus');
jest.mock('../../../../src/api.v2/helpers/pipeline/hooks/HookRunner');

jest.mock('../../../../src/utils/schema-validator');

const uploadToAWSPayload = require('../../mocks/data/gem2sUploadToAWSPayload.json');

const experimentInstance = Experiment();
const sampleInstance = Sample();
const experimentExecutionInstance = ExperimentExecution();

const hookRunnerInstance = HookRunner();

describe('createGem2sPipeline', () => {
  const experimentId = 'mockExperimentId';
  const paramsHash = 'mockParamsHash';
  const authJWT = 'mockAuthJWT';

  const mockExperiment = {
    id: '8e282f0d-aadb-8032-a334-982a371efd0f',
    name: 'asdsadsada',
    description: 'Analysis description',
    samplesOrder: ['fc68aefc-c3ca-467f-8589-f1dbaaac1c1e'],
    processingConfig: {},
    notifyByEmail: true,
    createdAt: '2022-05-10 15:41:04.165961+00',
    updatedAt: '2022-05-10 15:41:04.165961+00',
  };

  const mockSamples = [{
    id: 'fc68aefc-c3ca-467f-8589-f1dbaaac1c1e',
    experimentId: '8e282f0d-aadb-8032-a334-982a371efd0f',
    name: 'WT1',
    sampleTechnology: '10x',
    createdAt: '2022-05-10 15:41:10.057808+00',
    updatedAt: '2022-05-10 15:41:10.057808+00',
    metadata: { Track_1: 'N.A.' },
    files: {
      matrix10x: {
        size: 5079737, s3Path: '68f74995-3689-401a-90e0-145e08049cd5', uploadStatus: 'uploaded', sampleFileType: 'matrix10x',
      },
      barcodes10x: {
        size: 5331, s3Path: '37d9e601-9278-437c-a776-40fe94680833', uploadStatus: 'uploaded', sampleFileType: 'barcodes10x',
      },
      features10x: {
        size: 279361, s3Path: '1ee00087-e98a-4390-a3cb-392a3f6d09d8', uploadStatus: 'uploaded', sampleFileType: 'features10x',
      },
    },
  }];

  const mockStateMachineArn = 'mockStateMachineArn';
  const mockExecutionArn = 'mockExecutionArn';

  beforeEach(() => {
    experimentInstance.findById.mockClear();
    sampleInstance.getSamples.mockClear();
    experimentExecutionInstance.upsert.mockClear();
    pipelineConstruct.createGem2SPipeline.mockClear();

    experimentInstance.findById.mockReturnValueOnce({
      first: jest.fn(() => Promise.resolve(mockExperiment)),
    });

    sampleInstance.getSamples.mockReturnValueOnce(Promise.resolve(mockSamples));

    pipelineConstruct.createGem2SPipeline.mockReturnValueOnce(
      { stateMachineArn: mockStateMachineArn, executionArn: mockExecutionArn },
    );
  });

  it('works correctly', async () => {
    await createGem2sPipeline(experimentId, { paramsHash }, authJWT);

    expect(experimentInstance.findById).toHaveBeenCalledWith(experimentId);
    expect(sampleInstance.getSamples).toHaveBeenCalledWith(experimentId);
    expect(experimentExecutionInstance.upsert.mock.calls[0]).toMatchSnapshot();
    expect(pipelineConstruct.createGem2SPipeline.mock.calls[0]).toMatchSnapshot();
  });
});

const mockGetPipelineStatusResponse = {
  status: {
    gem2s: {
      startDate: '2022-05-10T16:30:16.268Z',
      stopDate: null,
      status: 'RUNNING',
      error: false,
      completedSteps: ['DownloadGem'],
      paramsHash: 'mockParamsHash',
    },
  },
};

describe('gem2sResponse', () => {
  const experimentId = 'mockExperimentId';
  const message = {
    experimentId, authJWT: 'mockAuthJWT', input: { authJWT: 'mockAuthJWT' }, response: {},
  };

  beforeEach(() => {
    io.sockets = { emit: jest.fn() };

    experimentInstance.updateById.mockClear();
    pipelineConstruct.createQCPipeline.mockClear();

    getPipelineStatus.mockReturnValueOnce(mockGetPipelineStatusResponse);
  });

  it('works correctly', async () => {
    await handleGem2sResponse(io, message);

    expect(validateRequest).toHaveBeenCalledWith(message, 'GEM2SResponse.v2.yaml');
    expect(hookRunnerInstance.run).toHaveBeenCalledWith(message);

    expect(getPipelineStatus).toHaveBeenCalledWith(experimentId, constants.GEM2S_PROCESS_NAME);
    expect(io.sockets.emit.mock.calls[0]).toMatchSnapshot();
  });

  it('Starts a QC run when gem2s finishes', async () => {
    const stateMachineArn = 'mockStateMachineArn';
    const executionArn = 'mockExecutionArn';

    pipelineConstruct.createQCPipeline.mockImplementationOnce(
      () => Promise.resolve({ stateMachineArn, executionArn }),
    );

    // There's a hook registered on the uploadToAWS step
    expect(hookRunnerInstance.register.mock.calls[0][0]).toEqual('uploadToAWS');

    const hookedFunctions = hookRunnerInstance.register.mock.calls[0][1];

    expect(hookedFunctions).toHaveLength(1);

    // calling the hookedFunction triggers QC
    await hookedFunctions[0](uploadToAWSPayload);

    expect(experimentInstance.updateById.mock.calls).toMatchSnapshot();

    expect(pipelineConstruct.createQCPipeline.mock.calls).toMatchSnapshot();
  });
});
