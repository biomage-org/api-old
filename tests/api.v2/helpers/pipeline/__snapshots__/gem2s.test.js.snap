// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createGem2sPipeline works correctly 1`] = `
Array [
  Object {
    "experiment_id": "mockExperimentId",
    "pipeline_type": "gem2s",
  },
  Object {
    "execution_arn": "mockExecutionArn",
    "params_hash": "mockParamsHash",
    "state_machine_arn": "mockStateMachineArn",
  },
]
`;

exports[`createGem2sPipeline works correctly 2`] = `
Array [
  Object {
    "experiment_id": "mockExperimentId",
    "pipeline_type": "qc",
  },
]
`;

exports[`createGem2sPipeline works correctly 3`] = `
Array [
  "mockExperimentId",
  Object {
    "authJWT": "mockAuthJWT",
    "experimentName": "asdsadsada",
    "input": Object {
      "type": "10x",
    },
    "metadata": Object {
      "Track_1": Array [
        "N.A.",
      ],
    },
    "organism": null,
    "projectId": "mockExperimentId",
    "sampleIds": Array [
      "fc68aefc-c3ca-467f-8589-f1dbaaac1c1e",
    ],
    "sampleNames": Array [
      "WT1",
    ],
    "sampleOptions": Object {
      "fc68aefc-c3ca-467f-8589-f1dbaaac1c1e": Object {},
    },
    "sampleS3Paths": Object {
      "fc68aefc-c3ca-467f-8589-f1dbaaac1c1e": Object {
        "barcodes10x": "37d9e601-9278-437c-a776-40fe94680833",
        "features10x": "1ee00087-e98a-4390-a3cb-392a3f6d09d8",
        "matrix10x": "68f74995-3689-401a-90e0-145e08049cd5",
      },
    },
  },
]
`;

exports[`gem2sResponse Starts a QC run when gem2s finishes 1`] = `
Array [
  Array [
    "06d50e28-cc96-2624-900c-e0e9a0995598",
    Object {
      "processing_config": Object {
        "cellSizeDistribution": Object {
          "auto": true,
          "enabled": false,
          "f87892f0-3403-4ba9-b871-c366e3fa855e": Object {
            "auto": true,
            "defaultFilterSettings": Object {
              "binStep": 200,
              "minCellSize": 1044,
            },
            "enabled": false,
            "filterSettings": Object {
              "binStep": 200,
              "minCellSize": 1044,
            },
          },
          "filterSettings": Object {
            "binStep": 200,
            "minCellSize": 1080,
          },
        },
        "classifier": Object {
          "auto": true,
          "enabled": false,
          "f87892f0-3403-4ba9-b871-c366e3fa855e": Object {
            "auto": true,
            "defaultFilterSettings": Object {
              "FDR": 0.01,
            },
            "enabled": false,
            "filterSettings": Object {
              "FDR": 0.01,
            },
          },
          "filterSettings": Object {
            "FDR": 0.01,
          },
          "prefiltered": true,
        },
        "configureEmbedding": Object {
          "clusteringSettings": Object {
            "method": "louvain",
            "methodSettings": Object {
              "louvain": Object {
                "resolution": 0.8,
              },
            },
          },
          "embeddingSettings": Object {
            "method": "umap",
            "methodSettings": Object {
              "tsne": Object {
                "learningRate": 200,
                "perplexity": 9.18,
              },
              "umap": Object {
                "distanceMetric": "cosine",
                "minimumDistance": 0.3,
              },
            },
          },
        },
        "dataIntegration": Object {
          "dataIntegration": Object {
            "method": "harmony",
            "methodSettings": Object {
              "fastmnn": Object {
                "normalisation": "logNormalize",
                "numGenes": 2000,
              },
              "harmony": Object {
                "normalisation": "logNormalize",
                "numGenes": 2000,
              },
              "seuratv4": Object {
                "normalisation": "logNormalize",
                "numGenes": 2000,
              },
              "unisample": Object {
                "normalisation": "logNormalize",
                "numGenes": 2000,
              },
            },
          },
          "dimensionalityReduction": Object {
            "excludeGeneCategories": Array [],
            "method": "rpca",
            "numPCs": null,
          },
        },
        "doubletScores": Object {
          "auto": true,
          "enabled": true,
          "f87892f0-3403-4ba9-b871-c366e3fa855e": Object {
            "auto": true,
            "defaultFilterSettings": Object {
              "binStep": 0.02,
              "probabilityThreshold": 0.7622447,
            },
            "enabled": true,
            "filterSettings": Object {
              "binStep": 0.02,
              "probabilityThreshold": 0.7622447,
            },
          },
          "filterSettings": Object {
            "binStep": 0.02,
            "probabilityThreshold": 0.5,
          },
        },
        "mitochondrialContent": Object {
          "auto": true,
          "enabled": true,
          "f87892f0-3403-4ba9-b871-c366e3fa855e": Object {
            "auto": true,
            "defaultFilterSettings": Object {
              "method": "absoluteThreshold",
              "methodSettings": Object {
                "absoluteThreshold": Object {
                  "binStep": 0.3,
                  "maxFraction": 0.07181329,
                },
              },
            },
            "filterSettings": Object {
              "method": "absoluteThreshold",
              "methodSettings": Object {
                "absoluteThreshold": Object {
                  "binStep": 0.3,
                  "maxFraction": 0.07181329,
                },
              },
            },
          },
          "filterSettings": Object {
            "method": "absoluteThreshold",
            "methodSettings": Object {
              "absoluteThreshold": Object {
                "binStep": 0.3,
                "maxFraction": 0.1,
              },
            },
          },
        },
        "numGenesVsNumUmis": Object {
          "auto": true,
          "enabled": true,
          "f87892f0-3403-4ba9-b871-c366e3fa855e": Object {
            "auto": true,
            "defaultFilterSettings": Object {
              "regressionType": "linear",
              "regressionTypeSettings": Object {
                "linear": Object {
                  "p.level": 0.001,
                },
                "spline": Object {
                  "p.level": 0.001,
                },
              },
            },
            "enabled": true,
            "filterSettings": Object {
              "regressionType": "linear",
              "regressionTypeSettings": Object {
                "linear": Object {
                  "p.level": 0.001,
                },
                "spline": Object {
                  "p.level": 0.001,
                },
              },
            },
          },
          "filterSettings": Object {
            "regressionType": "linear",
            "regressionTypeSettings": Object {
              "linear": Object {
                "p.level": 0.001,
              },
              "spline": Object {
                "p.level": 0.001,
              },
            },
          },
        },
      },
    },
  ],
]
`;

exports[`gem2sResponse Starts a QC run when gem2s finishes 2`] = `
Array [
  Array [
    "06d50e28-cc96-2624-900c-e0e9a0995598",
    Array [],
    "Bearer someHugeHash",
    "a6d49c90-74f0-4c1e-aa8f-673312b6eb7d",
  ],
]
`;

exports[`gem2sResponse works correctly 1`] = `
Array [
  "ExperimentUpdates-mockExperimentId",
  Object {
    "experimentId": "mockExperimentId",
    "input": Object {},
    "response": Object {},
    "status": Object {
      "status": Object {
        "gem2s": Object {
          "completedSteps": Array [
            "DownloadGem",
          ],
          "error": false,
          "paramsHash": "mockParamsHash",
          "startDate": "2022-05-10T16:30:16.268Z",
          "status": "RUNNING",
          "stopDate": null,
        },
      },
    },
    "type": "gem2s",
  },
]
`;
